### PR TITLE
feat: synchronously fetch coin balances when an address is viewed.

### LINKED ISSUE #1156: Check the coin balance when a user visits the address page
There may be situations where a value transfer occurs but we have not updated the coin balance for an address. When a visitor lands on the address page we should check the balance of that address immediately and update the balance in real-time.

### PR DESCRIPTION
**Pull request by zachdaniel**
Resolves #1156 	

### Enhancements
* Visiting the address page now ensures we have a reasonably recent balance (within the past estimated day of blocks).
* Show the block that we have most recently updated the balance for.
* update the diffing logic for new balances on the FE. It used to just check if we sent it a balance card with different html. However, the balance card we send up always says `Fetching tokens...`, so it will almost always be different from what is there, since we fetch the tokens right away. So this sends along the balance in wei, and it uses that to determine if this new balance should be rendered. This should result in less reloading of that component for no reason.
* When an address is visited, we check if it has a stale (not recently updated) balance. If so, we double-check it, and/or import any pending coin balances for it.
* When the address page is finished loading, it asks the address channel for the current coin balance. This helps w/ the race condition of the on-demand update finishing before the client has had a chance to join the channel.

This should probably wait on https://github.com/poanetwork/blockscout/issues/1379 because it needs to know the latest block. Once that is merged, I'll update this to use it.

### CODE REVIEW

File: apps/block_scout_web/test/block_scout_web/features/address_contract_verification_test.exs [hunk 1]
```diff
@@ -1,5 +1,5 @@
 defmodule BlockScoutWeb.AddressContractVerificationTest do
-  use BlockScoutWeb.FeatureCase, async: true
+  use BlockScoutWeb.FeatureCase, async: false
```
Comment (ayrat555) [inline apps/block_scout_web/test/block_scout_web/features/address_contract_verification_test.exs]: Why did you change it?

Comment (zachdaniel) [reply apps/block_scout_web/test/block_scout_web/features/address_contract_verification_test.exs]: I was having an issue I can't recall now. But I think I can undo it no problem.

Comment (zachdaniel) [reply apps/block_scout_web/test/block_scout_web/features/address_contract_verification_test.exs]: there is new work that happens asynchronously, so this is required to avoid errors.

**Comment by acravenho** [comment]
@zachdaniel I'm seeing the following error every few minutes:

```
2019-02-05T15:09:16.622 [error] GenServer BlockScoutWeb.RealtimeEventHandler terminating
** (RuntimeError) cannot encode metadata from the :__meta__ field for Explorer.Chain.Address to JSON. This metadata is used internally by ecto and should never be exposed externally.

You can either map the schemas to remove the :__meta__ field before encoding to JSON, or explicit list the JSON fields in your schema:

    defmodule Explorer.Chain.Address do
      # ...

      @derive {Poison.Encoder, only: [:name, :title, ...]}
      schema ... do

    (ecto) lib/ecto/json.ex:35: Poison.Encoder.Ecto.Schema.Metadata.encode/2
    (poison) lib/poison/encoder.ex:227: anonymous fn/4 in Poison.Encoder.Map.encode/3
    (poison) lib/poison/encoder.ex:228: Poison.Encoder.Map."-encode/3-lists^foldl/2-0-"/3
    (poison) lib/poison/encoder.ex:228: Poison.Encoder.Map.encode/3
    (poison) lib/poison/encoder.ex:227: anonymous fn/4 in Poison.Encoder.Map.encode/3
    (poison) lib/poison/encoder.ex:228: Poison.Encoder.Map."-encode/3-lists^foldl/2-0-"/3
    (poison) lib/poison/encoder.ex:228: Poison.Encoder.Map.encode/3
    (poison) lib/poison/encoder.ex:259: anonymous fn/3 in Poison.Encoder.List.encode/3
    (poison) lib/poison/encoder.ex:260: Poison.Encoder.List."-encode/3-lists^foldr/2-1-"/3
    (poison) lib/poison/encoder.ex:260: Poison.Encoder.List.encode/3
    (poison) lib/poison.ex:41: Poison.encode!/2
    (phoenix) lib/phoenix/socket/serializers/v2_json_serializer.ex:8: Phoenix.Socket.V2.JSONSerializer.fastlane!/1
    (phoenix) lib/phoenix/channel/server.ex:365: anonymous fn/5 in Phoenix.Channel.Server.fastlane/3
    (elixir) lib/enum.ex:1925: Enum."-reduce/3-lists^foldl/2-0-"/3
    (phoenix_pubsub) lib/phoenix/pubsub/local.ex:96: Phoenix.PubSub.Local.broadcast/6
    (elixir) lib/enum.ex:765: Enum."-each/2-lists^foreach/1-0-"/2
    (elixir) lib/enum.ex:765: Enum.each/2
    (phoenix_pubsub) lib/phoenix/pubsub/pg2_server.ex:30: Phoenix.PubSub.PG2Server.do_broadcast/7
    (block_scout_web) lib/block_scout_web/notifier.ex:176: BlockScoutWeb.Notifier.broadcast_transaction/3
    (elixir) lib/enum.ex:765: Enum."-each/2-lists^foreach/1-0-"/2
Last message: {:chain_event, :transactions, :realtime, [%Explorer.Chain.Transaction{v: nil, block_number: nil, nonce: nil, cumulative_gas_used: nil, from_address_hash: nil, to_address_hash: nil, r: nil, s: nil, created_contract_code_indexed_at: nil, from_address: #Ecto.Association.NotLoaded<association :from_address is not loaded>, value: nil, forks: #Ecto.Association.NotLoaded<association :forks is not loaded>, logs: #Ecto.Association.NotLoaded<association :logs is not loaded>, input: nil, error: nil, block_hash: nil, inserted_at: nil, token_transfers: #Ecto.Association.NotLoaded<association :token_transfers is not loaded>, hash: %Explorer.Chain.Hash{byte_count: 32, bytes: <<183, 10, 7, 139, 65, 203, 206, 184, 4, 112, 211, 200, 87, 69, 198, 117, 45, 115, 245, 206, 145, 111, 129, 128, ...>>}, internal_transactions: #Ecto.Association.NotLoaded<association :internal_transactions is not loaded>, __meta__: #Ecto.Schema.Metadata<:loaded, "transactions">, gas: nil, status: nil, to_address: #Ecto.Association.NotLoaded<association :to_address is not loaded>, uncles: #Ecto.Association.NotLoaded<association :uncles is not loaded>, gas_used: nil, created_contract_address: #Ecto.Association.NotLoaded<association :created_contract_address is not loaded>, gas_price: nil, internal_transactions_indexed_at: nil, index: nil, block: #Ecto.Association.NotLoaded<association :block is not loaded>, created_contract_address_hash: nil, updated_at: nil}]}
State: []
```

**Comment by zachdaniel** [comment]
Looking now.

**Comment by zachdaniel** [comment]
Hmm...that is interesting.I didn't really change any of the code around what gets broadcast, I just wrote code to trigger broadcasts.

**Comment by acravenho** [comment]
I was viewing `0x1fddEc96688e0538A316C64dcFd211c491ECf0d8` which updates every minute or so on POA Core and received the following error:

```
2019-02-06T12:51:56.808 [error] GenServer BlockScoutWeb.RealtimeEventHandler terminating
** (RuntimeError) cannot encode association :contracts_creation_internal_transaction from Explorer.Chain.Address to JSON because the association was not loaded.

You can either preload the association:

    Repo.preload(Explorer.Chain.Address, :contracts_creation_internal_transaction)

Or choose to not encode the association when converting the struct to JSON by explicitly listing the JSON fields in your schema:

    defmodule Explorer.Chain.Address do
      # ...

      @derive {Poison.Encoder, only: [:name, :title, ...]}
      schema ... do

    (ecto) lib/ecto/json.ex:13: Poison.Encoder.Ecto.Association.NotLoaded.encode/2
    (poison) lib/poison/encoder.ex:227: anonymous fn/4 in Poison.Encoder.Map.encode/3
    (poison) lib/poison/encoder.ex:228: Poison.Encoder.Map."-encode/3-lists^foldl/2-0-"/3
    (poison) lib/poison/encoder.ex:228: Poison.Encoder.Map.encode/3
    (poison) lib/poison/encoder.ex:227: anonymous fn/4 in Poison.Encoder.Map.encode/3
    (poison) lib/poison/encoder.ex:228: Poison.Encoder.Map."-encode/3-lists^foldl/2-0-"/3
    (poison) lib/poison/encoder.ex:228: Poison.Encoder.Map.encode/3
    (poison) lib/poison/encoder.ex:259: anonymous fn/3 in Poison.Encoder.List.encode/3
    (poison) lib/poison/encoder.ex:260: Poison.Encoder.List."-encode/3-lists^foldr/2-1-"/3
    (poison) lib/poison/encoder.ex:260: Poison.Encoder.List.encode/3
    (poison) lib/poison.ex:41: Poison.encode!/2
    (phoenix) lib/phoenix/socket/serializers/v2_json_serializer.ex:8: Phoenix.Socket.V2.JSONSerializer.fastlane!/1
    (phoenix) lib/phoenix/channel/server.ex:365: anonymous fn/5 in Phoenix.Channel.Server.fastlane/3
    (elixir) lib/enum.ex:1925: Enum."-reduce/3-lists^foldl/2-0-"/3
    (phoenix_pubsub) lib/phoenix/pubsub/local.ex:96: Phoenix.PubSub.Local.broadcast/6
    (elixir) lib/enum.ex:765: Enum."-each/2-lists^foreach/1-0-"/2
    (elixir) lib/enum.ex:765: Enum.each/2
    (phoenix_pubsub) lib/phoenix/pubsub/pg2_server.ex:30: Phoenix.PubSub.PG2Server.do_broadcast/7
    (block_scout_web) lib/block_scout_web/notifier.ex:176: BlockScoutWeb.Notifier.broadcast_transaction/3
    (elixir) lib/enum.ex:765: Enum."-each/2-lists^foreach/1-0-"/2
Last message: {:chain_event, :transactions, :realtime, [%Explorer.Chain.Transaction{block_number: nil, nonce: nil, token_transfers: #Ecto.Association.NotLoaded<association :token_transfers is not loaded>, v: nil, to_address: #Ecto.Association.NotLoaded<association :to_address is not loaded>, created_contract_code_indexed_at: nil, r: nil, s: nil, internal_transactions_indexed_at: nil, created_contract_address_hash: nil, from_address_hash: nil, value: nil, to_address_hash: nil, forks: #Ecto.Association.NotLoaded<association :forks is not loaded>, created_contract_address: #Ecto.Association.NotLoaded<association :created_contract_address is not loaded>, input: nil, from_address: #Ecto.Association.NotLoaded<association :from_address is not loaded>, error: nil, gas: nil, inserted_at: nil, hash: %Explorer.Chain.Hash{byte_count: 32, bytes: <<25, 246, 174, 144, 19, 128, 159, 178, 45, 203, 82, 27, 78, 79, 249, 34, 136, 98, 19, 120, 197, 213, ...>>}, __meta__: #Ecto.Schema.Metadata<:loaded, "transactions">, cumulative_gas_used: nil, status: nil, gas_price: nil, logs: #Ecto.Association.NotLoaded<association :logs is not loaded>, gas_used: nil, index: nil, block_hash: nil, block: #Ecto.Association.NotLoaded<association :block is not loaded>, uncles: #Ecto.Association.NotLoaded<association :uncles is not loaded>, internal_transactions: #Ecto.Association.NotLoaded<association :internal_transactions is not loaded>, updated_at: nil}]}
```

**Comment by acravenho** [comment]
I'm no longer seeing addresses update. I started indexing POA Core, navigated to an address that I knew had a balance. Waited several minutes, no update. I checked the logs to make sure the coin balance fetcher wasn't overloaded. 

This is balance I checked: `0x57Da3F24D81186B344649FE92B120AD45100D749`

**Comment by acravenho** [comment]
screen shot 2019-02-11 at 10 41 22 am

Let's move the Last Updated to the left side so the Balance back is not cluttered.

File: apps/block_scout_web/lib/block_scout_web/templates/address/overview.html.eex [hunk 2]
```diff
@@ -33,16 +33,28 @@
...
-                  <%= Cldr.Number.to_string!(@validation_count, format: "#,###") %>
-                </span> <%= gettext("Blocks Validated") %>
+                <span class="address-detail-item">
+                  <span data-selector="validation-count">
+                    <%= Cldr.Number.to_string!(@validation_count, format: "#,###") %>
+                  </span>
+                  <%= gettext("Blocks Validated") %>
+                </span>
               <% end %>
+              <span class="address-detail-item">1,000 Blocks Validated</span>
```
Comment (acravenho) [inline apps/block_scout_web/lib/block_scout_web/templates/address/overview.html.eex]: This line was accidentally copied over.

File: apps/block_scout_web/lib/block_scout_web/templates/address/overview.html.eex [hunk 3]
```diff
@@ -33,16 +33,28 @@
...
+              <span class="address-detail-item">
+                <span data-selector="transaction-count">
+                  <%= Cldr.Number.to_string!(@transaction_count, format: "#,###") %>
+                </span>
+                <%= gettext("Transactions Sent") %>
+              </span>
+              <span class="address-detail-item">
+                <%= gettext("Last Balance Update: Block #") %>
+                <span data-selector="fetched-coin-balance-block-number">
+                    <%= @address.fetched_coin_balance_block_number %>
```
Comment (acravenho) [inline apps/block_scout_web/lib/block_scout_web/templates/address/overview.html.eex]: Can we remove the space between the `#` and the actual block number?
